### PR TITLE
feat: support multiple pools

### PR DESCRIPTION
**Pull request by brianium**
There are a few chores in here for process.

This will probably be a 2.0 release as it changes the API quite a bit.

Notable changes:

* Drops Laravel provider - this can be handled pretty easily in the laravel-o-sphere and it's cumbersome to test this from this repo.
* ClaimsVerifiers now verify claims as a function of a `UserPool` type. This makes it easier to conceptually separate pools and clients
* JWKs are now a requirement of the `UserPool` and need to be passed as data instead of file paths
* There is a `MultiPoolJwtDecoder` that delegates to `JwtDecoder` but allows an array of verifiers to be passed.

### CODE REVIEW

**Comment by brianium** [comment]
@johncorrelli this PR adds pretty clean support for multiple cognito pools to juhwit. 

Sonar cloud is having a bad time with the duplication in tests. Can we turn that rule off for this repo? Or maybe ignore test directories?

Once this bad brother is merged, I'll tag it and release 2.0 so we can use it for great good 🙇

**Comment by brianium** [comment]
Also, commit by commit is probably best for this feller

**Comment by johncorrelli** [comment]
@brianium - sounds groovy - looking at this now. As for sonarcloud - I updated the exclusions to ignore the spec directory. Unfortunately, there is no way to re-run a test without a push, but that's fine, we can merge without that passing if need be.

File: composer.json [hunk 1]
```diff
@@ -6,7 +6,7 @@
     "authors": [
         {
             "name": "Brian Scaturro",
-            "email": "scaturrob@gmail.com"
+            "email": "brian@teamgantt.com.com"
```
Comment (johncorrelli) [inline composer.json]: `.com.com`

File: README.md [hunk 2]
```diff
@@ -12,21 +12,26 @@ The main service provided by Juhwit is the `JwtDecoder` which is composed with t
...
-$verifier = new CognitoClaimVerifier($clientIds, $poolId, $region);
+// we need some public keys in the form of a jwk (accessible via cognito)
+$jwk = json_decode(file_get_contents('path/to/jwk.json'), true);
+
+$pool = new UserPool($poolId, $clientIds, $region, $jwk);
+$verifier = new CognitoClaimVerifier($pool);
 $decoder = new JwtDecoder($verifier);
 
-// we need some public keys in the form of a jwk (accessible via cognito)
+
```
Comment (johncorrelli) [inline README.md]: if we're being picky - we don't need this extra whitespace here... but only if we're being picky.

File: README.md [hunk 3]
```diff
@@ -12,21 +12,26 @@ The main service provided by Juhwit is the `JwtDecoder` which is composed with t
...
 $pathToJwk = '/some/path/to/jwk.json';
 
 // If all is valid we will get a token back - otherwise a TokenException is thrown
-$token = $decoder->decode($someTokenFromARequest, $pathToJwk);
+$token = $decoder->decode($someTokenFromARequest);
```
Comment (johncorrelli) [inline README.md:31]: is `$pathToJwk` above no longer needed in this example? or do we need to add more example?

File: spec/MultiPoolJwtDecoder.spec.php [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+            $token = $decoder->decode($this->jwt);
+            expect($token->getClaim('custom:user_id'))->toBe('123');
+            expect($token->getClaim('token_use'))->toBe('id');
+        });
+
+        it('should report the last known error produced by a claims verifier', function () {
+            $first = Double::instance(['implements' => ClaimVerifierInterface::class]);
+        
+            allow($first)->toReceive('verify')->andRun(function ($token) {
+                throw new InvalidClaimsException("Those claims are not invalid!");
```
Comment (johncorrelli) [inline spec/MultiPoolJwtDecoder.spec.php]: I know this is just a test, but we have a double negative in the exception. Also, doesn't look like we're using token.

File: spec/MultiPoolJwtDecoder.spec.php [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+            });
+
+            allow($first)->toReceive('getUserPool')->andReturn($this->userPool);
+
+            $second = Double::instance(['implements' => ClaimVerifierInterface::class]);
+        
+            allow($second)->toReceive('verify')->andRun(function ($token) {
+                throw new InvalidClaimsException("Those claims are not invalid! Not even a little bit!");
```
Comment (johncorrelli) [inline spec/MultiPoolJwtDecoder.spec.php]: super not invalid. not even a lot :trollface: And `$token`

File: spec/MultiPoolJwtDecoder.spec.php [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+        it('should decode a valid jwt', function () {
+            $token = $this->decoder->decode($this->jwt);
+            expect($token->getClaim('custom:user_id'))->toBe('123');
+            expect($token->getClaim('token_use'))->toBe('id');
+        });
+
+        it('should accept valid claims if only one verifier succeeds', function () {
+            $invalidClaimsVerifier = Double::instance(['implements' => ClaimVerifierInterface::class]);
+        
+            allow($invalidClaimsVerifier)->toReceive('verify')->andRun(function ($token) {
```
Comment (johncorrelli) [inline spec/MultiPoolJwtDecoder.spec.php]: Looks like we're not using token here, is it needed?

**Review by johncorrelli — CHANGES_REQUESTED** [review]
Looks great! just a few small things. I think on the next push, it should fix the sonarcloud duplication stuff too.

**Review by johncorrelli — APPROVED** [review]

**Comment by johncorrelli** [comment]
:boom: Ready to merge! :shipit:

**Pull request merged**